File: backend/timezone_config.py
"""
Timezone Configuration for Backend

Gerencia configuração de timezone do sistema backend.
Usa variável de ambiente TZ para configurar timezone do Python e PostgreSQL.

Default: America/Sao_Paulo (UTC-3)

Para alterar, configure TZ no .env:
    TZ=America/Manaus  # UTC-4
    TZ=America/Rio_Branco  # UTC-5
"""

import os
import logging
from datetime import datetime, timezone
from zoneinfo import ZoneInfo
from typing import Optional

logger = logging.getLogger(__name__)

# Timezones suportados (Brasil)
SUPPORTED_TIMEZONES = {
    'America/Sao_Paulo': {'offset': -3, 'name': 'Horário de Brasília'},
    'America/Fortaleza': {'offset': -3, 'name': 'Fortaleza'},
    'America/Recife': {'offset': -3, 'name': 'Recife'},
    'America/Belem': {'offset': -3, 'name': 'Belém'},
    'America/Maceio': {'offset': -3, 'name': 'Maceió'},
    'America/Bahia': {'offset': -3, 'name': 'Salvador'},
    'America/Araguaina': {'offset': -3, 'name': 'Araguaína'},
    'America/Manaus': {'offset': -4, 'name': 'Manaus'},
    'America/Cuiaba': {'offset': -4, 'name': 'Cuiabá'},
    'America/Campo_Grande': {'offset': -4, 'name': 'Campo Grande'},
    'America/Boa_Vista': {'offset': -4, 'name': 'Boa Vista'},
    'America/Porto_Velho': {'offset': -4, 'name': 'Porto Velho'},
    'America/Rio_Branco': {'offset': -5, 'name': 'Rio Branco'},
    'America/Noronha': {'offset': -2, 'name': 'Fernando de Noronha'}
}

# Default timezone
DEFAULT_TIMEZONE = 'America/Sao_Paulo'


def get_timezone() -> str:
    """
    Obtém timezone configurado via variável de ambiente TZ

    Returns:
        str: Nome do timezone (ex: 'America/Sao_Paulo')
    """
    tz = os.getenv('TZ', DEFAULT_TIMEZONE)

    if tz not in SUPPORTED_TIMEZONES:
        logger.warning(
            f"Timezone '{tz}' não suportado. Usando default: {DEFAULT_TIMEZONE}"
        )
        return DEFAULT_TIMEZONE

    return tz


def get_timezone_info() -> dict:
    """
    Obtém informações sobre o timezone configurado

    Returns:
        dict: Informações do timezone (name, offset)
    """
    tz = get_timezone()
    info = SUPPORTED_TIMEZONES.get(tz, SUPPORTED_TIMEZONES[DEFAULT_TIMEZONE])

    return {
        'timezone': tz,
        'offset': info['offset'],
        'name': info['name']
    }


def get_timezone_obj() -> ZoneInfo:
    """
    Obtém objeto ZoneInfo do timezone configurado

    Returns:
        ZoneInfo: Objeto timezone para uso com datetime
    """
    return ZoneInfo(get_timezone())


def now_local() -> datetime:
    """
    Obtém datetime atual no timezone local configurado

    Returns:
        datetime: Datetime atual com timezone
    """
    return datetime.now(get_timezone_obj())


def to_local(dt: datetime) -> datetime:
    """
    Converte datetime UTC para timezone local

    Args:
        dt: Datetime em UTC

    Returns:
        datetime: Datetime no timezone local
    """
    if dt.tzinfo is None:
        # Se não tem timezone, assume UTC
        dt = dt.replace(tzinfo=timezone.utc)

    return dt.astimezone(get_timezone_obj())


def to_utc(dt: datetime) -> datetime:
    """
    Converte datetime local para UTC

    Args:
        dt: Datetime no timezone local

    Returns:
        datetime: Datetime em UTC
    """
    if dt.tzinfo is None:
        # Se não tem timezone, assume timezone local
        dt = dt.replace(tzinfo=get_timezone_obj())

    return dt.astimezone(timezone.utc)


def format_datetime_local(dt: datetime) -> str:
    """
    Formata datetime para string no formato brasileiro

    Args:
        dt: Datetime para formatar

    Returns:
        str: String formatada (DD/MM/YYYY HH:MM:SS)
    """
    local_dt = to_local(dt) if dt.tzinfo else dt
    return local_dt.strftime('%d/%m/%Y %H:%M:%S')


def format_date_local(dt: datetime) -> str:
    """
    Formata data para string no formato brasileiro

    Args:
        dt: Datetime/Date para formatar

    Returns:
        str: String formatada (DD/MM/YYYY)
    """
    local_dt = to_local(dt) if dt.tzinfo else dt
    return local_dt.strftime('%d/%m/%Y')


def parse_date_br(date_str: str) -> datetime:
    """
    Parse string de data brasileira para datetime

    Args:
        date_str: String no formato DD/MM/YYYY

    Returns:
        datetime: Datetime no timezone local
    """
    dt = datetime.strptime(date_str, '%d/%m/%Y')
    return dt.replace(tzinfo=get_timezone_obj())


def parse_datetime_br(datetime_str: str) -> datetime:
    """
    Parse string de datetime brasileira para datetime

    Args:
        datetime_str: String no formato DD/MM/YYYY HH:MM:SS

    Returns:
        datetime: Datetime no timezone local
    """
    dt = datetime.strptime(datetime_str, '%d/%m/%Y %H:%M:%S')
    return dt.replace(tzinfo=get_timezone_obj())


# Log configuração no import
tz_info = get_timezone_info()
logger.info(
    f"🌍 Timezone configurado: {tz_info['name']} "
    f"({tz_info['timezone']}, UTC{tz_info['offset']:+d})"
)


# Exportar configuração para ser usada em queries SQL
def get_postgres_timezone_setting() -> str:
    """
    Retorna comando SQL para configurar timezone do PostgreSQL

    Returns:
        str: Comando SQL SET TIMEZONE
    """
    return f"SET TIMEZONE = '{get_timezone()}';"


# Para uso em configuração do PostgreSQL
POSTGRES_TIMEZONE = get_timezone()

File: src/config/timezone.config.ts
/**
 * Configuração de Timezone
 *
 * Centraliza a configuração de timezone do sistema
 * Pode ser controlado por variável de ambiente VITE_TIMEZONE
 *
 * Timezones suportados:
 * - America/Sao_Paulo (UTC-3) - Padrão
 * - America/Fortaleza (UTC-3)
 * - America/Recife (UTC-3)
 * - America/Manaus (UTC-4)
 * - America/Cuiaba (UTC-4)
 * - America/Campo_Grande (UTC-4)
 * - America/Belem (UTC-3)
 * - America/Maceio (UTC-3)
 * - America/Bahia (UTC-3)
 * - America/Araguaina (UTC-3)
 * - America/Noronha (UTC-2)
 * - America/Boa_Vista (UTC-4)
 * - America/Porto_Velho (UTC-4)
 * - America/Rio_Branco (UTC-5)
 */

export interface TimezoneConfig {
  /** Nome do timezone (ex: America/Sao_Paulo) */
  timezone: string;
  /** Offset em relação ao UTC (ex: -3, -4) */
  utcOffset: number;
  /** Código de localidade (ex: pt-BR) */
  locale: string;
  /** Nome amigável do timezone */
  displayName: string;
}

/**
 * Mapeamento de timezones brasileiros
 */
const TIMEZONE_MAP: Record<string, TimezoneConfig> = {
  'America/Sao_Paulo': {
    timezone: 'America/Sao_Paulo',
    utcOffset: -3,
    locale: 'pt-BR',
    displayName: 'Horário de Brasília (UTC-3)'
  },
  'America/Fortaleza': {
    timezone: 'America/Fortaleza',
    utcOffset: -3,
    locale: 'pt-BR',
    displayName: 'Fortaleza (UTC-3)'
  },
  'America/Recife': {
    timezone: 'America/Recife',
    utcOffset: -3,
    locale: 'pt-BR',
    displayName: 'Recife (UTC-3)'
  },
  'America/Manaus': {
    timezone: 'America/Manaus',
    utcOffset: -4,
    locale: 'pt-BR',
    displayName: 'Manaus (UTC-4)'
  },
  'America/Cuiaba': {
    timezone: 'America/Cuiaba',
    utcOffset: -4,
    locale: 'pt-BR',
    displayName: 'Cuiabá (UTC-4)'
  },
  'America/Campo_Grande': {
    timezone: 'America/Campo_Grande',
    utcOffset: -4,
    locale: 'pt-BR',
    displayName: 'Campo Grande (UTC-4)'
  },
  'America/Belem': {
    timezone: 'America/Belem',
    utcOffset: -3,
    locale: 'pt-BR',
    displayName: 'Belém (UTC-3)'
  },
  'America/Maceio': {
    timezone: 'America/Maceio',
    utcOffset: -3,
    locale: 'pt-BR',
    displayName: 'Maceió (UTC-3)'
  },
  'America/Bahia': {
    timezone: 'America/Bahia',
    utcOffset: -3,
    locale: 'pt-BR',
    displayName: 'Salvador (UTC-3)'
  },
  'America/Araguaina': {
    timezone: 'America/Araguaina',
    utcOffset: -3,
    locale: 'pt-BR',
    displayName: 'Araguaína (UTC-3)'
  },
  'America/Noronha': {
    timezone: 'America/Noronha',
    utcOffset: -2,
    locale: 'pt-BR',
    displayName: 'Fernando de Noronha (UTC-2)'
  },
  'America/Boa_Vista': {
    timezone: 'America/Boa_Vista',
    utcOffset: -4,
    locale: 'pt-BR',
    displayName: 'Boa Vista (UTC-4)'
  },
  'America/Porto_Velho': {
    timezone: 'America/Porto_Velho',
    utcOffset: -4,
    locale: 'pt-BR',
    displayName: 'Porto Velho (UTC-4)'
  },
  'America/Rio_Branco': {
    timezone: 'America/Rio_Branco',
    utcOffset: -5,
    locale: 'pt-BR',
    displayName: 'Rio Branco (UTC-5)'
  }
};

/**
 * Obtém o timezone configurado
 * Prioridade:
 * 1. Variável de ambiente VITE_TIMEZONE
 * 2. Valor padrão: America/Sao_Paulo
 */
export const getTimezone = (): string => {
  const envTimezone = import.meta.env.VITE_TIMEZONE;

  if (envTimezone && TIMEZONE_MAP[envTimezone]) {
    console.log(`✅ Timezone configurado via env: ${envTimezone}`);
    return envTimezone;
  }

  // Fallback para São Paulo
  const defaultTimezone = 'America/Sao_Paulo';
  console.log(`⚙️ Usando timezone padrão: ${defaultTimezone}`);
  return defaultTimezone;
};

/**
 * Obtém a configuração completa do timezone atual
 */
export const getTimezoneConfig = (): TimezoneConfig => {
  const timezone = getTimezone();
  return TIMEZONE_MAP[timezone] || TIMEZONE_MAP['America/Sao_Paulo'];
};

/**
 * Obtém o offset UTC do timezone atual
 */
export const getTimezoneOffset = (): number => {
  return getTimezoneConfig().utcOffset;
};

/**
 * Obtém o locale do timezone atual
 */
export const getLocale = (): string => {
  return getTimezoneConfig().locale;
};

/**
 * Lista todos os timezones disponíveis
 */
export const getAvailableTimezones = (): TimezoneConfig[] => {
  return Object.values(TIMEZONE_MAP);
};

/**
 * Valida se um timezone é suportado
 */
export const isValidTimezone = (timezone: string): boolean => {
  return timezone in TIMEZONE_MAP;
};

/**
 * Aplica o offset do timezone a uma data
 * NOTA: Esta função é para casos especiais onde conversão de timezone é necessária
 * Para a maioria dos casos, use as funções de dateUtils.ts que NÃO convertem timezone
 */
export const applyTimezoneOffset = (date: Date, targetTimezone?: string): Date => {
  const tz = targetTimezone || getTimezone();
  const config = TIMEZONE_MAP[tz];

  if (!config) {
    console.warn(`Timezone ${tz} não encontrado, usando São Paulo`);
    return date;
  }

  // Calcula o offset em milissegundos
  const offsetMs = config.utcOffset * 60 * 60 * 1000;

  // Aplica o offset
  return new Date(date.getTime() + offsetMs);
};

// Log da configuração atual no console (apenas em desenvolvimento)
if (import.meta.env.DEV) {
  const config = getTimezoneConfig();
  console.log('🌍 Configuração de Timezone:', {
    timezone: config.timezone,
    utcOffset: config.utcOffset,
    locale: config.locale,
    displayName: config.displayName
  });
}

export default {
  getTimezone,
  getTimezoneConfig,
  getTimezoneOffset,
  getLocale,
  getAvailableTimezones,
  isValidTimezone,
  applyTimezoneOffset
};
